File: packages/react-native-webcrypto-bridge/android/src/main/java/com/okta/webcryptobridge/CryptoAlgorithmHandler.kt
package com.okta.webcryptobridge

import android.security.keystore.KeyGenParameterSpec
import org.json.JSONObject
import java.security.PublicKey

/**
 * Interface for algorithm-specific cryptographic operations.
 *
 * Implementations of this interface encapsulate all algorithm-specific logic for key generation,
 * import/export, and signing operations. This allows the main WebCryptoBridgeModule to remain
 * algorithm-agnostic and makes it easy to add support for new algorithms (EC, EdDSA) without
 * modifying the core module logic.
 */
interface CryptoAlgorithmHandler {
    /**
     * Generates a KeyGenParameterSpec for this algorithm.
     *
     * This spec is used to configure the Android KeyPairGenerator for key generation in the
     * Android Keystore. The handler is responsible for validating algorithm parameters and
     * throwing IllegalArgumentException if the parameters are invalid (e.g., unsupported key size).
     *
     * @param alias the keystore alias for the key being generated
     * @param params JSON object containing algorithm-specific parameters (e.g., `modulusLength` for RSA)
     * @param purposes bit flags indicating key usage (e.g., KeyProperties.PURPOSE_SIGN)
     * @return a configured KeyGenParameterSpec for this algorithm
     * @throws IllegalArgumentException if algorithm parameters are invalid
     */
    fun generateKeySpec(alias: String, params: JSONObject, purposes: Int): KeyGenParameterSpec

    /**
     * Exports a public key to JWK (JSON Web Key) format.
     *
     * Handles algorithm-specific JWK fields. For example:
     * - RSA exports the modulus (`n`) and public exponent (`e`)
     * - EC would export the curve point coordinates (`x`, `y`)
     *
     * @param publicKey the public key to export
     * @return a JSONObject containing the JWK representation (always includes `kty` and `alg`)
     */
    fun exportToJWK(publicKey: PublicKey): JSONObject

    /**
     * Imports a public key from JWK (JSON Web Key) format.
     *
     * Handles algorithm-specific key reconstruction from JWK fields. For example:
     * - RSA reconstructs the key from modulus (`n`) and public exponent (`e`)
     * - EC would reconstruct from curve point coordinates (`x`, `y`)
     *
     * @param jwk the JWK object containing algorithm-specific fields
     * @return the reconstructed PublicKey
     * @throws IllegalArgumentException if JWK is malformed or contains invalid values
     */
    fun importFromJWK(jwk: JSONObject): PublicKey

    /**
     * Returns the signature algorithm string for this algorithm.
     *
     * This string is passed to `Signature.getInstance()` to obtain the appropriate
     * signature provider. Examples:
     * - `"SHA256withRSA"` for RSA with SHA-256
     * - `"SHA256withECDSA"` for ECDSA with SHA-256
     *
     * @return the signature algorithm string (e.g., `"SHA256withRSA"`)
     */
    fun getSignatureAlgorithm(): String
}

File: packages/react-native-webcrypto-bridge/ios/CryptoAlgorithmHandler.swift
import Security

/// Specification for key generation, returned by algorithm handlers.
struct KeyGenSpec {
    /// SecKeyAttribute key type (e.g., kSecAttrKeyTypeRSA)
    let keyType: CFString
    /// Key size in bits
    let keySize: Int
}

/// Protocol for algorithm-specific cryptographic operations.
/// Implementations handle key generation specs, JWK export/import, and signature algorithm selection
/// for specific algorithms (RSA, ECDSA, EdDSA, etc).
protocol CryptoAlgorithmHandler {
    /// Generates key generation specifications for this algorithm.
    /// Validates algorithm parameters and returns the specs needed for SecKeyCreateRandomKey.
    ///
    /// @param params Dictionary containing algorithm parameters (e.g., "modulusLength" for RSA)
    /// @return KeyGenSpec with keyType and keySize
    /// @throws NSError if parameters are invalid or unsupported for this algorithm
    func generateKeySpec(_ params: [String: Any]) throws -> KeyGenSpec

    /// Returns the SecKeyAlgorithm constant for this algorithm's signing/verification operations.
    /// @return SecKeyAlgorithm (e.g., .rsaSignatureMessagePKCS1v15SHA256 for RSA)
    func getSignatureAlgorithm() -> SecKeyAlgorithm

    /// Exports a public key to JWK (JSON Web Key) format.
    /// Handles algorithm-specific JWK fields (e.g., "n", "e" for RSA; "x", "y" for EC).
    ///
    /// @param publicKey The SecKey to export
    /// @param keyComponents RSAPublicKeyComponents containing modulus and exponent
    /// @return Dictionary with JWK representation including "kty", algorithm identifier, and key components
    func exportToJWK(publicKey: SecKey, keyComponents: RSAPublicKeyComponents) -> [String: Any]

    /// Imports a public key from JWK (JSON Web Key) format.
    /// Validates JWK structure and extracts algorithm-specific components.
    ///
    /// @param jwk Dictionary representation of a JWK
    /// @return RSAPublicKeyComponents if parsing succeeds, nil otherwise
    func importFromJWK(_ jwk: [String: Any]) -> RSAPublicKeyComponents?
}
